solver/llbsolver/solver.go

```diff
@@ -176,55 +176,40 @@ func (s *Solver) Solve(ctx context.Context, id string, sessionID string, req fro
 			return nil, err
 		}
 		if len(dtbi) > 0 {
-			if res.Metadata == nil {
-				res.Metadata = make(map[string][]byte)
-			}
-			res.Metadata[exptypes.ExporterBuildInfo] = dtbi
+			res.AddMeta(exptypes.ExporterBuildInfo, dtbi)
 		}
 	}
-	if res.Refs != nil {
-		for k, r := range res.Refs {
-			if r == nil {
-				continue
-			}
-			dtbi, err := buildinfo.Encode(ctx, res.Metadata, fmt.Sprintf("%s/%s", exptypes.ExporterBuildInfo, k), r.BuildSources())
-			if err != nil {
-				return nil, err
-			}
-			if len(dtbi) > 0 {
-				if res.Metadata == nil {
-					res.Metadata = make(map[string][]byte)
-				}
-				res.Metadata[fmt.Sprintf("%s/%s", exptypes.ExporterBuildInfo, k)] = dtbi
-			}
+	for k, r := range res.Refs {
+		if r == nil {
+			continue
+		}
+		dtbi, err := buildinfo.Encode(ctx, res.Metadata, fmt.Sprintf("%s/%s", exptypes.ExporterBuildInfo, k), r.BuildSources())
+		if err != nil {
+			return nil, err
+		}
+		if len(dtbi) > 0 {
+			res.AddMeta(fmt.Sprintf("%s/%s", exptypes.ExporterBuildInfo, k), dtbi)
 		}
 	}
 
 	var exporterResponse map[string]string
 	if e := exp.Exporter; e != nil {
-		inp := exporter.Source{
-			Metadata:     res.Metadata,
-			Attestations: res.Attestations,
-		}
-		if inp.Metadata == nil {
-			inp.Metadata = make(map[string][]byte)
-		}
-		if inp.Attestations == nil {
-			inp.Attestations = make(map[string][]attestation.Attestation)
+		cached, err := result.ConvertResult(res, func(res solver.ResultProxy) (solver.CachedResult, error) {
+			return res.Result(ctx)
+		})
+		if err != nil {
+			return nil, err
 		}
-		var cr solver.CachedResult
-		var crMap = map[string]solver.CachedResult{}
-		if res := res.Ref; res != nil {
-			r, err := res.Result(ctx)
-			if err != nil {
-				return nil, err
-			}
-			workerRef, ok := r.Sys().(*worker.WorkerRef)
+
+		inp, err := result.ConvertResult(cached, func(res solver.CachedResult) (cache.ImmutableRef, error) {
+			workerRef, ok := res.Sys().(*worker.WorkerRef)
 			if !ok {
-				return nil, errors.Errorf("invalid reference: %T", r.Sys())
+				return nil, errors.Errorf("invalid reference: %T", res.Sys())
 			}
-			inp.Ref = workerRef.ImmutableRef
-			cr = r
+			return workerRef.ImmutableRef, nil
+		})
+		if err != nil {
+			return nil, err
 		}
 		if res.Refs != nil {
```

Comment:
Woops, just noticed this little section during review of this, this block isn't needed anymore - I've opened a PR to clear it up in https://github.com/moby/buildkit/pull/3086, apologies :smile: